### PR TITLE
Pass invalidation handler to ExtensionProcess::grantCapability

### PR DESCRIPTION
**Pull request by pvollan**
#### 11913004d540d2edc9ab32dffa2e1c15624e73cb
<pre>
Pass invalidation handler to ExtensionProcess::grantCapability
<a href="https://bugs.webkit.org/show_bug.cgi?id=274331">https://bugs.webkit.org/show_bug.cgi?id=274331</a>
<a href="https://rdar.apple.com/126825362">rdar://126825362</a>

Reviewed by Chris Dumez.

In order for the invalidation handler to be called when the grant is invalidated, we need to
pass the invalidation handler to ExtensionProcess::grantCapability.

* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::Function&lt;void):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
(WebKit::ExtensionProcess::grantCapability const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::LaunchGrant::LaunchGrant):
(WebKit::LaunchGrant::~LaunchGrant):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: <a href="https://commits.webkit.org/279428@main">https://commits.webkit.org/279428@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a4e26e49a8c6d718895594043b762867b877ffe6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56760 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4206 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/40300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3988 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43346 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2753 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55578 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31021 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46215 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24482 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2362 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58355 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28637 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3704 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46411 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30772 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7870 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29614 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm [hunk 1]
```diff
@@ -388,14 +388,6 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
     if (process.extensionProcess()) {
         ASCIILiteral runningBoardAssertionName = runningBoardNameForAssertionType(m_assertionType);
         ASCIILiteral runningBoardDomain = runningBoardDomainForAssertionType(m_assertionType);
-        auto didInvalidateBlock = [weakThis = ThreadSafeWeakPtr { *this }, runningBoardAssertionName] () {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm]: Why this change?

Comment (bnham) [reply Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm]: Presumably so the code can be shared with the call on L477

File: Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm [hunk 2]
```diff
@@ -84,7 +88,7 @@
     return grant;
 }
 
-PlatformGrant ExtensionProcess::grantCapability(const PlatformCapability& capability) const
+PlatformGrant ExtensionProcess::grantCapability(const PlatformCapability& capability, Function<void()>&& invalidationHandler) const
```
Comment (bnham) [inline Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm]: Could this take an `AssertionCapability` instead of a `PlatformCapability` as an argument and just use the `m_didInvalidateBlock` from `AssertionCapability`? It seems a bit weird to be passing the invalidation block in two different places...

File: Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm (hunk 1 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm]: Yes, that is correct.

Thanks for reviewing, all!

File: Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm (hunk 2 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm]: That's a good point! I will make that change.

Comment (pvollan) [reply Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm]: I started on this change, but it became more involving than expected, because this method is also used to grant Media capabilities. Would you be OK with the current approach, Ben?

**Comment by pvollan** [comment]
Looking closer, I think the test failures on iOS are unrelated to this patch.

**Review by cdumez — APPROVED** [review]

**Comment by cdumez** [comment]
> Pass invalidation handler to ExtensionProcess::grantCapability
> https://bugs.webkit.org/show_bug.cgi?id=274331
> [rdar://126825362](https://rdar.apple.com/126825362)

This radar is already in Integrate so I doubt this is the right link.

**Comment by pvollan** [comment]
> > Pass invalidation handler to ExtensionProcess::grantCapability
> > https://bugs.webkit.org/show_bug.cgi?id=274331
> > [rdar://126825362](https://rdar.apple.com/126825362)
> 
> This radar is already in Integrate so I doubt this is the right link.

Ah, good catch, the radar state was not correct.

Thanks for approving, Chris!
